### PR TITLE
Merge from devel_lp to devel

### PR DESCRIPTION
**Pull request by ericahan22**
This pull request implements fanout inflation based on linear programming. It also adds the possibility of hybrid sources. Both InflationLP and InflationSDP use the same monomial class.

### CODE REVIEW

File: inflation/lp/InflationLP.py [hunk 1]
```diff
@@ -0,0 +1,1890 @@
...
+        self._lexorder = inflationproblem._lexorder
+        self._nr_operators = inflationproblem._nr_operators
+        self._lexrange = np.arange(self._nr_operators)
+        self.lexorder_symmetries = inflationproblem.lexorder_symmetries
+        self._from_2dndarray = inflationproblem._from_2dndarray
+        self.mon_to_lexrepr = inflationproblem.mon_to_lexrepr
+        self.blank_bool_vec = np.zeros(self._nr_operators, dtype=bool)
+        self._ortho_groups_per_party = inflationproblem._ortho_groups_per_party
+        self.has_children = inflationproblem.has_children.copy()
+        if include_all_outcomes or supports_problem: # HACK to fix detection of incompatible supports. (Can be fixed upon adding set_extra_equalities)
```
Comment (ericahan22) [inline inflation/lp/InflationLP.py:115]: Has this been fixed?

File: inflation/lp/InflationLP.py [hunk 2]
```diff
@@ -0,0 +1,1890 @@
...
+
+        self.identity_operator = np.empty((0, self._nr_properties),
+                                          dtype=self.np_dtype)
+        self.zero_operator = np.zeros((1, self._nr_properties),
+                                      dtype=self.np_dtype)
+
+        # These next properties are reset during generate_lp, but are needed in
+        # init so as to be able to test the Monomial constructor function
+        # without generate_lp.
+        #TODO: Enable test of Monomial constructor without generate_lp!
```
Comment (ericahan22) [inline inflation/lp/InflationLP.py:173]: Has this been done?

File: inflation/lp/InflationLP.py [hunk 3]
```diff
@@ -0,0 +1,1890 @@
...
+            return self.monomial_from_atoms[key]
+        except KeyError:
+            mon = CompoundMoment(atoms)
+            try:
+                mon.idx = self.first_free_idx
+                self.first_free_idx += 1
+            except AttributeError:
+                pass
+            self.monomial_from_atoms[key] = mon
+            self.monomial_from_name[mon.name] = mon  # TODO: Add simple name and complex name
```
Comment (ericahan22) [inline inflation/lp/InflationLP.py:980]: Have we done this/does this need to be done?

File: inflation/lp/InflationLP.py [hunk 4]
```diff
@@ -0,0 +1,1890 @@
...
+            List of dictionaries.
+        """
+        input_lil_mat = input_coo_mat.tolil(copy=False)
+        args_iter = zip(input_lil_mat.rows, input_lil_mat.data)
+        if string_keys:
+            return [self._coo_vec_to_name_dict(*args) for args in args_iter]
+        else:
+            return [self._coo_vec_to_mon_dict(*args) for args in args_iter]
+
+    def _mon_dict_to_coo_vec(self, monomials_dict: Dict) -> coo_matrix:
```
Comment (ericahan22) [inline inflation/lp/InflationLP.py:1416]: This function is still being used, confirm that we are not deprecating it in this merge?

Comment (ericahan22) [reply inflation/lp/InflationLP.py:1416]: See sparse_objective, sparse_known_vars, sparse_lowerbounds, sparse_upperbounds

File: inflation/lp/InflationLP.py [hunk 5]
```diff
@@ -0,0 +1,1890 @@
...
+        nonfanout : bool, optional
+            Whether to consider only nonfanout inflation (GPT problem) or
+            fanout inflation (classical problem). By default ``False``.
+        supports_problem : bool, optional
+            Whether to consider feasibility problems with distributions, or 
+            just with the distribution's support. By default ``False``.
+        default_non_negative : bool, optional
+            Whether to set all variables to be non-negative by default. By
+            default ``True``.
+        include_all_outcomes : bool, optional
```
Comment (ericahan22) [inline inflation/lp/InflationLP.py:64]: Confirm that this is what include_all_outcomes actually means

File: inflation/lp/InflationLP.py [hunk 6]
```diff
@@ -0,0 +1,1890 @@
...
+        """List of dictionaries of moment equalities."""
+        return self._coo_mat_to_dict(self.sparse_equalities)
+
+    @property
+    def moment_inequalities(self):
+        """List of dictionaries of moment inequalities."""
+        return self._coo_mat_to_dict(self.sparse_inequalities)
+
+
+    def _discover_inflation_orbits(self, _raw_monomials_as_lexboolvecs) -> np.ndarray:
```
Comment (ericahan22) [inline inflation/lp/InflationLP.py:1438]: Type of _raw_monomials_as_lexboolvecs?

File: test/test_pipeline.py [hunk 7]
```diff
@@ -720,3 +850,268 @@ def test_normeqs_colineq2momentineq(self):
...
+                            "identify the incompatible distribution.")
+        with self.subTest(msg="Testing GHZ, compatible distribution"):
+            lp.set_distribution(GHZ(crit_cutoff - 1e-2))
+            lp.solve()
+            self.assertEqual(lp.status, "optimal",
+                             "The LP did not recognize the compatible "
+                             "distribution.")
+        with self.subTest(msg="Testing GHZ, compatible distribution, "
+                              "feasibility as optimization"):
+            self.skipTest("Feasibility as optimization not working for "
```
Comment (ericahan22) [inline test/test_pipeline.py:901]: Check if this is fixed or not

File: inflation/lp/InflationLP.py [hunk 8]
```diff
@@ -0,0 +1,1862 @@
...
+                                 reversed(choices_to_combine_CG))
+        raw_boolvecs_global = reduce(nb_outer_bitwise_or,
+                                     reversed(choices_to_combine_global))
+        return (raw_boolvecs_CG[np.argsort(raw_boolvecs_CG.sum(axis=1))],
+                raw_boolvecs_global[np.argsort(np.matmul(raw_boolvecs_global,
+                                                         self.boolvec_for_CG_ineqs.astype(int))
+                                               )])
+
+    @cached_property
+    def sparse_equalities_OLD(self) -> coo_matrix:
```
Comment (ecboghiu) [inline inflation/lp/InflationLP.py:1158]: @eliewolfe remove this?
```suggestion
    def sparse_equalities_OLD(self) -> coo_matrix:
```

File: inflation/lp/InflationLP.py [hunk 9]
```diff
@@ -0,0 +1,1862 @@
...
+        self.atomic_monomial_from_hash  = dict()
+        self.monomial_from_atoms        = dict()
+        self.monomial_from_name         = dict()
+        self.One  = self.Monomial(self.identity_operator, idx=1)
+        self._generate_lp()
+
+    ###########################################################################
+    # MAIN ROUTINES EXPOSED TO THE USER                                       #
+    ###########################################################################
+    def _generate_lp(self) -> None:
```
Comment (ecboghiu) [inline inflation/lp/InflationLP.py:182]: @Elie Wolfe rename to generate_relaxation() ?
```suggestion
    def _generate_lp(self) -> None:
```

File: test/test_pipeline.py (hunk 7 above)
Comment (ecboghiu) [reply test/test_pipeline.py:901]: New commit changing >=0 to >=-1e-6.

File: inflation/lp/InflationLP.py (hunk 1 above)
Comment (ecboghiu) [reply inflation/lp/InflationLP.py:114]: Openes a new issue to keep track of this.

File: inflation/lp/InflationLP.py (hunk 3 above)
Comment (ecboghiu) [reply inflation/lp/InflationLP.py:1027]: Added a new issue.

File: inflation/lp/InflationLP.py (hunk 2 above)
Comment (ecboghiu) [reply inflation/lp/InflationLP.py:172]: Added an issue to keep track of this.

File: inflation/lp/InflationLP.py (hunk 5 above)
Comment (ecboghiu) [reply inflation/lp/InflationLP.py:63]: It is correct.

File: inflation/lp/InflationLP.py (hunk 6 above)
Comment (ecboghiu) [reply inflation/lp/InflationLP.py:1473]: Added type

File: inflation/lp/InflationLP.py (hunk 4 above)
Comment (ecboghiu) [reply inflation/lp/InflationLP.py:1451]: We are not deprecating this function yet, so it's ok.

**Pull request merged**